### PR TITLE
Add device selection to `et run`

### PR DESCRIPTION
**Pull request by johnmccutchan**
- `et run` now detects the target device automatically and builds the related engine builds.
- `et run -- -d <device-id>` also works.

### CODE REVIEW

File: tools/engine_tool/lib/src/json_utils.dart [hunk 1]
```diff
@@ -0,0 +1,75 @@
+void _appendTypeError(
```
Comment (zanderso) [inline tools/engine_tool/lib/src/json_utils.dart:1]: Needs the engine copyright header.

File: tools/engine_tool/lib/src/commands/run_command.dart [hunk 2]
```diff
@@ -76,26 +77,47 @@ final class RunCommand extends CommandBase {
     return null;
   }
 
-  String _selectTargetConfig() {
-    final String configName = argResults![configFlag] as String;
-    if (configName.isNotEmpty) {
-      return configName;
+  String _getDeviceId() {
```
Comment (zanderso) [inline tools/engine_tool/lib/src/commands/run_command.dart:80]: There are a couple of ways to test this. Maybe the simplest is to add tests like the `RunCommand` tests that already exist, and observe the `flutter` commands that are spawned as a result. Another option is to make these methods public, and unit test them by adding `@visibleForTesting`.

File: tools/engine_tool/lib/src/run_utils.dart [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    }
+  }
+  return defaultDevice(env, targets);
+}
+
+/// Detects available targets and then selects one.
+Future<RunTarget?> detectAndSelectRunTarget(Environment env,
+    [String? idPrefix]) async {
+  final ProcessRunnerResult result = await env.processRunner
+      .runProcess(<String>['flutter', 'devices', '--machine']);
```
Comment (zanderso) [inline tools/engine_tool/lib/src/run_utils.dart:122]: Not for this PR, but something to keep in mind: At some point we'll have a need to explicitly specify the path to the Flutter SDK.

File: tools/engine_tool/lib/src/json_utils.dart (hunk 1 above)
Comment (johnmccutchan) [reply tools/engine_tool/lib/src/json_utils.dart:1]: Done.

File: tools/engine_tool/lib/src/run_utils.dart (hunk 3 above)
Comment (johnmccutchan) [reply tools/engine_tool/lib/src/run_utils.dart:122]: ACK.

File: tools/engine_tool/lib/src/commands/run_command.dart (hunk 2 above)
Comment (johnmccutchan) [reply tools/engine_tool/lib/src/commands/run_command.dart:80]: Added a test.

**Review by zanderso — APPROVED** [review]

**Pull request merged**